### PR TITLE
RWMutex initialisation fix & improved logging 

### PR DESCRIPTION
**Pull request by SebastianSchildt**
RWMutex tests failed on OS X if the memory used by "pthread_rwlock_t" used in pthread_rwlock_init looks like a locked pthread_rwlock_t structure.

This is fixed by initialising the lock structure using memset. PTHREAD_RWLOCK_INITIALIZER seems to be implemented differently on different systems, and therefore cannot be used on Linux and OS X in the same way (also not libcs might support it yet).

### CODE REVIEW

File: ibrcommon/ibrcommon/thread/RWMutex.cpp [hunk 1]
```diff
@@ -21,17 +21,58 @@
...
 #include "ibrcommon/thread/RWMutex.h"
 #include <errno.h>
+#include <cstring>
 
 namespace ibrcommon
 {
 	RWMutex::RWMutex()
 	{
-		pthread_rwlock_init(&_rwlock, NULL);
+		std::memset(&_rwlock,0,sizeof(pthread_rwlock_t));
```
Comment (morgenroth) [inline ibrcommon/ibrcommon/thread/RWMutex.cpp]: memset belongs to the default namespace. Hence, the right call would be ::memset(...).

File: ibrcommon/ibrcommon/thread/RWMutex.cpp [hunk 2]
```diff
@@ -21,17 +21,58 @@
...
+		
+		switch(pthread_rwlock_init(&_rwlock, NULL))
```
Comment (morgenroth) [inline ibrcommon/ibrcommon/thread/RWMutex.cpp]: This is bad practice.
Exceptions thrown by constructors are hard to catch. This is especially true if the objects are initialized via an initialization list.

File: ibrcommon/ibrcommon/thread/RWMutex.cpp [hunk 3]
```diff
@@ -21,17 +21,58 @@
...
 	RWMutex::~RWMutex()
 	{
-		pthread_rwlock_destroy(&_rwlock);
+		switch(pthread_rwlock_destroy(&_rwlock))
+		{
+			case 0:
+				break;
+				
+			case EPERM:
+				throw MutexException("The mutex could not be destroyed because the caller does not have sufficient privileges (EPERM).");
```
Comment (morgenroth) [inline ibrcommon/ibrcommon/thread/RWMutex.cpp]: This is a even more bad practice.
Exceptions thrown by destructors are impossible to catch and leaves the object in a undefined state.

File: ibrcommon/tests/thread/MutexTests.cpp [hunk 4]
```diff
@@ -103,6 +103,9 @@ void MutexTests::rwmutex_test_readonly()
 
 	// try to lock the mutex for a read-write access (should fail)
 	CPPUNIT_ASSERT_THROW(rwm.trylock_wr(), ibrcommon::MutexException);
+	
+	//Release original lock
+	CPPUNIT_ASSERT_NO_THROW(rwm.leave());
```
Comment (morgenroth) [inline ibrcommon/tests/thread/MutexTests.cpp]: leave() is already called by the destructor of 'l'. Calling the method a second time may lead to undefined behavior.

**Comment by SebastianSchildt** [comment]
Exceptions: Basically agree. In constructors they are fine (not always convenient to catch), in destructors they are evil. However: In this case both failure modes are fatal, i.e. the result should be a termination of the program. In the (now fixed) constructor the only reason for failing is that the system is not being able to provide the lock -> It is not possible to continue in any meaningful way. Failure in the destructor means, the program logic leading to the destructor had bugs such as forgetting to free locks, and the program should terminate (no way to "fix" this in the destructor). Question is, is there any other standard way of terminating a program using ibrcommon than throwing uncatched exceptions? 

(Need to look at the testcase again. Not sure why the RWLock unlock did not work on this machine without the fix.  However this is sort of a special case: The unittest triggers a (normally fatal) flaw, and needs to continue with the testcase. Could probably "#ifdef DARWIN" the hack in the unittest as so far it seems to be the only platform we have that is doing sanity checks on pthread_rwlock_destroy)

**Pull request closed without merging**